---
title: "04 velocyto with batch removal Jaeger/Jessberger"
params:
  seed: 1
author: "Izaskun Mallona (Mark Robinson lab)"
output:
  html_document:
    keep_md: false
    toc: true
    toc_float: true
    toc_depth: 4

---


# Context

# Load


```{r}
NTHREADS=20
```

```{r tag}
analysis_tag <- '04_velocyto_batch'
```

```{r libraries, include=TRUE, cache = FALSE}
library(pheatmap)
library(knitr)
library(scater)
library(Seurat)
library(Cairo)
library(edgeR)
library(data.table)
## library(biomaRt)
library(dplyr)
## library(reshape2)
#library(iSEE)
## library(Matrix)
## library(scran)
## library(mvoutlier)
## library(shiny)
## library(kableExtra)
## library(velocyto.R)
## library(cowplot)
## library(corrplot)
library(DT)
library(devtools)
## library(viridis)
## library("CellMixS")
library(gplots)
library(pheatmap)
library(velocyto.R)
```

```{r knitr_setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, comment = NULL, warning = TRUE,
                      fig.height = 8,
                      fig.width = 8,
                      cache = TRUE,
                      cache.lazy = FALSE,
                      tidy = TRUE, tidy.opts = list(width.cutoff=60))

options(bitmapType="cairo")

```

```{r functions}
ac <-  function(col, alpha=1){

    apply(sapply(col, col2rgb)/255, 2, 
          function(x) 
              rgb(x[1], x[2], x[3], alpha=alpha))  
}
```

# Data load

Provided by Jaeger

```{r}
so <- readRDS(file.path('..', 'data', 'Seurat_object_Gli_Ascl_combi_meta.Robj'))
```

Plus velocyto looms (from task `02_velocyto_mapping`)

```{r}

looms <- list()

for (fn in list.files(file.path('..', 'data', 'looms'), pattern = '.*loom', recursive = TRUE)) {

    ldat <- read.loom.matrices(file.path('..', 'data', 'looms', fn))

    ldat <- lapply(ldat,function(x) {
        colnames(x) <-  gsub("Aligned.sortedByCoord.out.bam","",colnames(x))
        x
    })

    ldat <- lapply(ldat,function(x) {
        colnames(x) <-  gsub(".*:","",colnames(x))
        x
    })

    looms[[dirname(fn)]] <- ldat
}
    

```

# Getting genes that are batch-specific

To be filtered out from velocyto calculations

```{r}
alt_batch <- as.character(so@meta.data$batch)


alt_batch[alt_batch == 'Ascl1_12wk_1'] <- 'B1'
alt_batch[alt_batch == 'Ascl1_12wk_2'] <- 'B2'
alt_batch[alt_batch == 'Ascl1_5day_1'] <- 'B1'
alt_batch[alt_batch == 'Ascl1_5day_2'] <- 'B2'
alt_batch[alt_batch == 'Gli1_12wk']    <- 'B1'
alt_batch[alt_batch == 'Gli1_5day_1']  <- 'B1'
alt_batch[alt_batch == 'Gli1_5day_2']  <- 'B1'
alt_batch <- as.factor(alt_batch)
table(alt_batch)
so@meta.data$alt_batch <- alt_batch


DefaultAssay(so) <- "RNA"
Idents(so) <- 'alt_batch'

markers <- FindMarkers(so,
                       ident.1 =  'B1',
                       ident.2 = 'B2',
                       test.use = "LR",
                       latent.vars = "combi",
                       min.pct = 0.2)

markers <- markers[order(markers$p_val_adj),]
markers <- markers[markers$p_val_adj < 0.05,]
markers$analysis <- 'batch_relevant_lr'
markers <- data.frame(gene = rownames(markers),
                      markers)

DT::datatable(markers %>% as.data.frame() %>% dplyr::mutate_if(is.numeric, 
    funs(round(., 2))), extensions = c("Buttons", "FixedColumns"), 
    rownames = FALSE, options = list(dom = "Bfrtip", scrollX = TRUE, 
        fixedColumns = list(leftColumns = 1), buttons = c("csv", "excel")))


```

## Subset loom files removing batch-affected genes

Removal of batch-affected genes (simple set differences between count tables and the significantly batch-affected genes)

```{r}
##selected <- setdiff(markers$gene, VariableFeatures(so))

for (loom in names(looms)) {
    for (item in names(looms[[loom]])) {
        curr <- setdiff(rownames(looms[[loom]][[item]]), markers$gene)
        looms[[loom]][[item]] <- looms[[loom]][[item]][curr,]
    }
}


```

Pasting loom files together (assumes equivalent gene sorting)

```{r}
ldat <- list()

ldat$spliced <- do.call(cbind.data.frame, sapply(looms, function(x) return (x$spliced)))
ldat$unspliced <- do.call(cbind.data.frame, sapply(looms, function(x) return (x$unspliced)))
ldat$spanning <- do.call(cbind.data.frame, sapply(looms, function(x) return (x$spanning)))

for (item in names(ldat)) {
    colnames(ldat[[item]]) <- sapply(strsplit(colnames(ldat[[item]]), '.', fixed = TRUE),
                                     function(x) return(x[2]))
}
```

Making sure we don't have the batch-affected genes.

```{r}

stopifnot(length(intersect(as.character(markers$gene), rownames(ldat$spliced))) == 0)

```

Remove cells that are not in use (Seurat's object contains the QC-ed cells only).

```{r}
for (item in names(ldat)) {
    ldat[[item]] <- ldat[[item]][,intersect(colnames(ldat[[item]]), colnames(so))]
}

```

Coloring cells according to the `combi` Jaeger's naming that includes the genotype and the cell cluster

```{r, cache = FALSE}
cell_colors <- so@meta.data['combi']
```

Distances as in Seurat's PCA

```{r}
cell_dist <- as.dist(1-armaCor(t(Embeddings(so, reduction = "pca"))))
```

```{r}
# exonic read (spliced) expression matrix
emat <- ldat$spliced;
# intronic read (unspliced) expression matrix
nmat <- ldat$unspliced
# spanning read (intron+exon) expression matrix
smat <- ldat$spanning;

summary(apply(emat,1,mean))
summary(apply(nmat,1,mean))
summary(apply(smat,1,mean))

```


```{r}
table(rownames(cell_colors) %in% colnames(emat))
## cell_colors <- data.frame(cell_colors[rownames(cell_colors) %in% colnames(emat),],
##                           row.names = rownames(cell_colors)[rownames(cell_colors)
##                                                            %in% colnames(emat)])

cell_colors$cell <- rownames(cell_colors)

cell_colors$color <- rainbow(8)[as.numeric(as.factor(cell_colors$combi))]

cnames <- cell_colors$cell

cell_colors <- cell_colors$color
names(cell_colors) <- cnames
```

Please mind and/or finetune the min max cluster average values

```{r, eval = TRUE, cache = FALSE}

## filter expression matrices based on some minimum max-cluster averages
emat <- filter.genes.by.cluster.expression(emat, cell_colors, min.max.cluster.average = 0.2)
nmat <- filter.genes.by.cluster.expression(nmat, cell_colors, min.max.cluster.average = 0.05)
smat <- filter.genes.by.cluster.expression(smat, cell_colors, min.max.cluster.average = 0.01)
# look at the resulting gene set
length(intersect(rownames(emat), rownames(nmat)))
```

Estimating velocities, note the k parameter


```{r, eval = TRUE, cache = FALSE}
set.seed(2)
fit.quantile <- 0.02;
rvel.qf <- gene.relative.velocity.estimates(as.matrix(emat),
                                            as.matrix(nmat),
                                            deltaT=1,
                                            kCells = 25,
                                            cell.dist = cell_dist,
                                            n.cores = NTHREADS,
                                            fit.quantile = fit.quantile)

pca.velocity.plot(rvel.qf, nPcs=5, plot.cols=2,
                  cell.colors=ac(cell_colors, alpha=0.7),
                  cex=1.2, pcount=0.1,
                  pc.multipliers=c(1, -1, -1, -1, -1))
```


Notice the number of neighbors (n= 100 etc)

```{r, eval = TRUE, cache = FALSE}
umap_coord <- Embeddings(so, reduction = "umap")

show.velocity.on.embedding.cor(
    umap_coord, rvel.qf, n=100, 
    scale='sqrt', cell.colors=ac(cell_colors, alpha=0.5), 
    cex=1, arrow.scale=4, arrow.lwd=1, do.par = TRUE,
    cell.border.alpha = 0.1)

```

```{r}

umap_coord_ascl = umap_coord[grep("Ascl", rownames(umap_coord)),]
umap_coord_gli = umap_coord[!rownames(umap_coord) %in% rownames(umap_coord_ascl),]
umap_coord_A_5d_12w_2 = umap_coord[grep("Ascl1_12wk_2|II", rownames(umap_coord)),]
```

## Velocities on subsets of cells

Ascl

```{r}
## Velocity for Ascl and Gli only
show.velocity.on.embedding.cor(umap_coord_ascl, rvel.qf, n = 100,
                               scale = 'sqrt',
                               cell.colors = ac(cell_colors, alpha = 0.5),
                               cex = 1, arrow.scale = 4, arrow.lwd = 1, do.par = T, cell.border.alpha  =  0.1)
```

A_5d_12w_2

```{r}
show.velocity.on.embedding.cor(umap_coord_A_5d_12w_2, rvel.qf, n = 100,
                               scale = 'sqrt', cell.colors = ac(cell_colors, alpha = 0.5),
                               cex = 1, arrow.scale = 4, arrow.lwd = 1, do.par = T, cell.border.alpha  =  0.1)
```

Gli

```{r}

show.velocity.on.embedding.cor(umap_coord_gli, rvel.qf, n = 100, scale = 'sqrt',
                               cell.colors = ac(cell_colors, alpha = 0.5), cex = 1,
                               arrow.scale = 4, arrow.lwd = 1, do.par = T, cell.border.alpha  =  0.1)
```

Shall we increase the number of neighbors while plotting? Example with 300 neighbors.

```{r, eval = TRUE, cache = FALSE}

show.velocity.on.embedding.cor(
    umap_coord, rvel.qf, n=300, 
    scale='sqrt', cell.colors=ac(cell_colors, alpha=0.5), 
    cex=1, arrow.scale = 4, arrow.lwd=1, do.par = TRUE,
    cell.border.alpha = 0.1)

```

# Save an image with all objects

```{r}
save.image('04_velocyto_batch_removal.RData')
```

# Session

```{r sessionInfo, cache = 0}

date()
devtools::session_info()


```
